scanpy/_utils/__init__.py

```diff
@@ -497,38 +497,29 @@ def get_ufuncs(data: np.ndarray | DaskArray):
     return np
 
 
-def check_nonnegative_integers(X: _FlatSupportedArray) -> bool:
-    """Checks values of X to ensure it is count data"""
+def check_nonnegative_integers(X: _SupportedArray) -> np.bool_ | DaskArray:
+    """\
+    Checks values of X to ensure it is count data.
+
+    When passed a dask array, it will return a scalar dask “array” evaluating to a boolean.
+    """
     from numbers import Integral
 
     ufuncs = get_ufuncs(X)
-    data = _get_values(X)
-    # Check no negatives
-    if ufuncs.signbit(data).any():
-        return False
-    # Check all are integers
-    if issubclass(data.dtype.type, Integral):
-        return True
-    # Check all are whole numbers
-    if ((data % 1) != 0).any():
-        return False
-    return True
-
-
-@singledispatch
-def _get_values(X: _FlatSupportedArray) -> np._SupportsArray:
-    raise NotImplementedError()
-
-
-@_get_values.register(np.ndarray)
-@_get_values.register(DaskArray)
-def _(X: np.ndarray) -> np._SupportsArray:
-    return X  # Doesn’t need to be flat
-
-
-@_get_values.register(sparse.spmatrix)
-def _(X: sparse.spmatrix) -> np._SupportsArray:
-    return X.data
+    # get (possibly nonzero) values as (possibly flat) array
+    data: np.ndarray | DaskArray = X.data if sparse.issparse(X) else X
+    # return bool or lazy dask array resolving to one
+    return (
+        # Check no negatives
+        (~ufuncs.signbit(data).any())
+        & (
+            # Check all are integers
+            issubclass(data.dtype.type, Integral)
+            |
```

Comment:
Won't this execute every check?

```python
In [1]: def foo():
   ...:     print("hi")
   ...:     return True
   ...: 

In [2]: True and foo()
hi
Out[2]: True

In [3]: True & foo()
hi
Out[3]: True
```

I think I'd prefer to keep as is so we exit as fast as possible.

Similarly, I wonder if we should just do a `da.map_blocks(check_nonnegative_integers).all()` to [avoid a large graph](https://docs.dask.org/en/stable/best-practices.html#avoid-very-large-graphs) for the dask case.

Reply:
OK, I did it! And it’s not even super ugly to write.